sky/provision/vast/utils.py

```
@@ -0,0 +1,153 @@
+
+    launch_params = {
+        'id': instance_touse['id'],
+        'direct': True,
+        'ssh': True,
+        'env': '-e __SOURCE=skypilot',
+        'onstart_cmd': ';'.join([
+            'touch ~/.no_auto_tmux',
+            f'echo "{vast.vast().api_key_access}" > ~/.vast_api_key',
+        ]),
+        'label': name,
+        'image': image_name
+    }
+
+    if preemptible:
+        launch_params['min_bid'] = instance_touse['min_bid']
+
+    new_instance_contract = vast.vast().create_instance(**launch_params)
+
+    new_instance = vast.vast().show_instance(
+        id=new_instance_contract['new_contract'])
+
+    return new_instance['id']
+
+
+def start(instance_id: str) -> None:
+    """Starts the given instance."""
+    vast.vast().start_instance(id=instance_id)
+
+
+def stop(instance_id: str) -> None:
+    """Stops the given instance."""
+    vast.vast().stop_instance(id=instance_id)
+
+
+def remove(instance_id: str) -> None:
+    """Terminates the given instance."""
+    vast.vast().destroy_instance(id=instance_id)
+
+
+def get_ssh_ports(cluster_name) -> List[int]:
+    """Gets the SSH ports for the given cluster."""
+    logger.debug(f'Getting SSH ports for cluster {cluster_name}.')
+
+    instances = list_instances()
+    possible_names = [f'{cluster_name}-head', f'{cluster_name}-worker']
+
+    ssh_ports = []
+
+    for instance in instances.values():
+        if instance['name'] in possible_names:
+            ssh_ports.append((instance['name'], instance['ssh_port']))
+    assert ssh_ports, (
+        f'Could not find any instances for cluster {cluster_name}.')
+
+    # So now we have
+    # [(name, port) ... ]
+    #
+    # We want to put head first and otherwise sort numerically
+    # and then extract the ports.
+    ssh_ports = list(
+        x[1]
+        for x in sorted(ssh_ports,
+                        key=lambda x: -1
+                        if x[0].endswith('head') else int(x[0].split('-')[-1])))
```

I'm a little bit confused about this. iiuc `x[0].split('-')[-1]` is guaranteed to be `worker` here? and this should throw an error?

Is this script works well with multi-node cluster?

---

we currently don't support num_nodes > 1 on vast

---

well I didn't just write that part for entertainment but the tests are passing without it, so whatever., here you go: /f83d11429d18e1ef56cca75531b775d2991435a4

---

alright this fails the format.sh test now ... I knew it was there for a reason ... let me look into it more.

---

alright: 82802162cca1baac6e118644689a418f5c334ec5